### DISCUSSION TITLE
EF Core query

### DISCUSSION THREAD

**Discussion by RichardD2** [LuckyPennySoftware/AutoMapper#4021] [NA]
I have an existing .NET Framework 4.8 WebAPI project using AutoMapper v10.1.1 and EF Core 3.1, which I am upgrading to .NET 6, AutoMapper 11.0.1, and EF Core 6.0.6.

The source model contains an owned entity:

```csharp
public class Settings
{
    public int Id { get; set; }
    ...
    public AdditionalSettings Additional { get; set; }
}

[Owned]
public class AdditionalSettings
{
    public bool Foo { get; set; }
    ...
}

...

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    base.OnModelCreating(modelBuilder);
    modelBuilder.Entity<Settings>().OwnsOne(s => s.Additional);
}
```

```sql
CREATE TABLE Settings
(
    Id int IDENTITY(1, 1) NOT NULL,
    ...
    Additional_Foo bit NULL,
    ...
);
```

The target model flattens the data:

```csharp
public class SettingsDto
{
    public int Id { get; set; }
    public bool Foo { get; set; }
}

...

configuration.CreateMap<Settings, SettingsDto>()
    .ForMember(s => s.Foo, o => o.MapFrom(s => s.Additional.Foo));
```

I then use `ProjectTo` to load a specific DTO instance:

```csharp
SettingsDto result = context.Set<Settings>()
    .ProjectTo<SettingsDto>(configurationProvider)
   .First(s => s.Id == 42);
```

In the .NET Framework version, this works as expected. If none of the owned entity columns are set, the `SettingsDto.Foo` property is set to `false`.

In the .NET 6 version, if none of the owned entity columns are set, I get an exception:

> Nullable object must have a value

To resolve the problem, I had to cast the `MapFrom` property to `bool?`, as mentioned on [StackOverflow](https://stackoverflow.com/a/32779419/124386):

```csharp
configuration.CreateMap<Settings, SettingsDto>()
    .ForMember(s => s.Foo, o => o.MapFrom(s => (bool?)s.Additional.Foo));
```

I couldn't see this change in behaviour documented anywhere. And I'm not entirely sure whether it's AutoMapper or EF Core which has caused it.

**Comment by lbargaoanu** [comment]
Check [the execution plan](https://docs.automapper.org/en/latest/Understanding-your-mapping.html). Compare it bettween the two versions. If it's the same, it's EF Core.

**Comment by RichardD2** [comment]
The generated queries are different. However, I suspect that's more due to improvements in EF Core.

## .NET Framework + EF Core 3.1 + AutoMapper 10.1:
```sql
SELECT TOP(1) [c].[Id], ..., [t4].[Foo] AS [Foo]
FROM [dbo].[Settings] AS [c]
LEFT JOIN (
    SELECT [c5].[Id], [c5].[Foo]
    FROM [dbo].[Settings] AS [c5]
    WHERE [c5].[Foo] IS NOT NULL
) AS [t4] ON [c].[Id] = [t4].[Id]
WHERE [c].[Id] = @__id_0
```
## .NET 6 + EF Core 6.0.6 + AutoMapper 11.0.1:
### Original query:
```sql
SELECT TOP(1) [c].[Id], ..., [c].[Foo] AS [Foo]
FROM [dbo].[Settings] AS [c]
WHERE [c].[Id] = @__id_0
```

### Fixed query:
```sql
SELECT TOP(1) [c].[Id], ... COALESCE([c].[Foo], CAST(0 AS bit)) AS [Foo]
FROM [dbo].[Settings] AS [c]
WHERE [c].[Id] = @__id_0
```

**Reply by jbogard** [reply]
Try checking the actual LINQ `Select` query between those versions with AutoMapper, you can recreate that Select query yourself too.

**Reply by RichardD2** [reply]
The LINQ expressions are the same for the .NET Framework and the unfixed .NET 6 queries. Pulling out the relevant part:

### .NET Framework + EF Core 3.1 + AutoMapper 10.1:
```
.Select(dtoSettings => new SettingsDto() { ... Foo = dtoSettings.Additional.Foo }).Where(c => c.Id == 11)
```

### .NET 6 + EF Core 6.0.6 + AutoMapper 11.0.1:
Original:
```
.Select(dtoSettings => new SettingsDto() { ... Foo = dtoSettings.Additional.Foo }).Where(c => c.Id == 11)
```

Fixed:
```
.Select(dtoSettings => new SettingsDto() { ... Foo = (Convert(dtoSettings.Additional.Foo, Nullable`1) ?? new Boolean()) }).Where(c => c.Id == 11)
```
